components/calendar/src/astronomy.rs

```diff
@@ -0,0 +1,572 @@
+    longitude: f32, // longitude from -180 to 180
+    elevation: f32, // elevation in meters
+}
+
+// The mean synodic month in days of 86400 atomic seconds
+// (86400 = 24 * 60 * 60)
+const MEAN_SYNODIC_MONTH: f64 = 29.530588853;
+
+// The Moment of noon on January 1, 2000
+const J2000: Moment = Moment::new(730120.5);
+
+impl Location {
+    /// Create a location; latitude is from -90 to 90, and longitude is from -180 to 180
+    pub fn new(latitude_input: f32, longitude_input: f32, elevation_input: f32) -> Location {
+        let result = Location {
+            latitude: latitude_input,
+            longitude: longitude_input,
+            elevation: elevation_input,
+        };
+        result.check();
+        result
+    }
+
+    pub fn check(&self) {
+        if self.latitude < -90.0 || self.latitude > 90.0 {
+            debug_assert!(false, "Locations must have latitudes from -90.0 to 90.0.");
+        }
+        if self.longitude < -180.00 || self.longitude > 180.0 {
+            debug_assert!(false, "Locations must have longitudes from -180 to 180.");
+        }
+    }
+
+    /// Get the direction from loc1 to loc2 in degrees east of due north
+    pub fn direction(loc1: Location, loc2: Location) -> f32 {
+        let y = (loc2.longitude - loc1.longitude).to_radians().sin();
+        let x = (loc1.latitude.to_radians().cos() * loc2.latitude.to_radians().tan())
+            - (loc1.latitude.to_radians().sin()
+                * (loc1.longitude - loc2.longitude).to_radians().cos());
+        if (x == y && y == 0.0) || loc2.latitude == 90.0 {
+            0.0
+        } else if loc2.latitude == -90.0 {
+            180.0
+        } else {
+            y.atan2(x)
+        }
+    }
+
+    /// Convert a longitude into a mean time zone;
+    /// this yields the difference in Moment given a longitude
+    /// e.g. a longitude of 90 degrees is 0.25 (90 / 360) days ahead
+    /// of a location with a longitude of 0 degrees.
+    pub fn zone_from_longitude(longitude: f64) -> f64 {
+        longitude / (360 as f64)
+    }
+
+    /// Convert from local mean time to universal time given a location
+    pub fn universal_from_local(local_time: Moment, location: Location) -> Moment {
+        local_time - Self::zone_from_longitude(location.longitude as f64)
+    }
+
+    /// Convert from universal time to local time given a location
+    pub fn local_from_universal(universal_time: Moment, location: Location) -> Moment {
+        universal_time + Self::zone_from_longitude(location.longitude as f64)
+    }
+}
+
+pub struct Astronomical;
+
+impl Astronomical {
+    /// Function for the ephemeris correction, which corrects the
+    /// somewhat-unpredictable discrepancy between dynamical time
+    /// and universal time
+    ///
+    /// Reference code: https://github.com/EdReingold/calendar-code2/blob/main/calendar.l#L3884-L3952
+    pub fn ephemeris_correction(moment: Moment) -> f64 {
+        let year: f64 = moment.inner().floor() / 364.2425;
```

Comment:
Issue: Pretty much anywhere we have `/` we need to use `div_euclid` instead

Reply:
I changed this in the latest commit, but this results in rounding issues which causes things to become 0 when they should be fractional values.